### DISCUSSION TITLE
Pathways for contributing/designing new graphics

### DISCUSSION THREAD

**Discussion by 12people** [CorsixTH/CorsixTH#2524] [Research Room]
It'd be useful if the README let designers know how to contribute, pointing them to the [Replacement Graphics repo](https://github.com/CorsixTH/CorsixTH-Graphics), letting them know how to submit graphics, and giving them guidelines on what these graphics should be like.

As for the guidelines, I would recommend taking inspiration from [OpenTTD](https://wiki.openttd.org/Graphics_Replacement#Contributer_Guidelines):

> New art is being drawn in the style of the original game, using the original 8bpp palette. The graphics should be a similar but distinct version of the object in question - no graphics may be copied at all from the original. The new sprites do not have to be the same size as the original, but need to be similar so as to fit into the game as expected. Where possible glitches/bugs in the original graphics should be avoided, for example the original snow line transition was in steps, whereas the new graphics give a smooth transition.

**Comment by TheCycoONE** [comment]
Part of the problem is that we don't know how we want to do that. The project stalled because the animations of people walking and using stuff: See discussion on https://github.com/CorsixTH/CorsixTH/pull/1138

**Comment by 12people** [comment]
@TheCycoONE The simplest solution here IMHO is to first focus on simpler objects that don't have animations, like plants, bins, walls, GUI, as mentioned in #1138. When those are redesigned, focus can shift to making animations of people and a technical solution for that can be devised then.

And if someone wants to work on these animations earlier, a developer can consult with this volunteer animator and find a solution that works for them.

**Comment by lewri** [comment]
Just going to copy this here from @Alberth289346 to be directly within the discussion (arguably you could move this into a discussion -- but notifications don't seem to be working 100% yet everywhere.

> 
> 
> Reading this again after 4 years makes the path more clear.
> ### Idea
> 
> I think what we want is
> a) Compatibility with the original graphics, and
> b) freedom to construct new graphics as we see fit for current animations.
> [c) freedom to add new animations, but that seems quite far away]
> 
> To make this happen, an animation should be self-supporting, it should not need any information about the animation to play from the program other than some position on the map.. That means all layer, mirror, and transparency flags should be included in the animation itself.
> 
> The game should only need to say "this animation please, at this spot in the map". A name or a unique number of the animation should be enough.
> (Note that frame-number as unique identity is likely to fail unless you decouple from the actual frame number, animations with new graphics may be ordered differently, and/or may be missing some animations.)
> 
> One way to achieve this is to add a "setupdata" block to each animation, which contains all the layer/mirror/transparency/etc information, as well as a link to the animation frames, colour palette, and what else you need. For the original graphics setupdata blocks cannot be added, so the loader would need to create those blocks (or load them from file?!), and attach the original animations to them.
> 
> Likely the setupdata should be extended to the point that we can show all patients, at least the radiation patient is missing, possibly others as well??
> 
> Such a setupdata block could have different content for new graphics, allowing a designer to change the internal structure of the animation, eg not share or layer the animation.
> 
> In the code you will get more animations, since flags etc are now encoded into different animations. In theory that can be solved by creating a table with flag settings to the animation-name to use.
> ### Supporting change in graphics mid-game
> 
> While experimenting with new graphics, Sadger at one point loaded an old savegame, while having a new plant animation available.
> The result was that all existing plants in the hospital were using the old (original) graphics, while newly bought plants used the new graphics.
> 
> The reason is that the save system saves and loads the running animations as well as all its settings in the various objects. A running animation doesn't query the animation frames to use from the animation collection on load.
> 
> While this makes for great fun, I would say this is undesired. When loading new graphics, the user should see new graphics when loading a game if we support that.
> This means that the save system should neither save running animations with its full details nor should it save any data with respect to which animations are currently available (things like the first frame number or number of frames etc). On a next load that data may be invalid, leading to anything from weird graphics to crashes.
> 
> Instead, it should query the name of the running animations while saving, as well as how far it progressed, so on the next load, a different animation can be loaded and positioned at a compatible point in time.
> This implies an animation needs a notion of progress that works between different graphics sets.
> ### New graphics
> 
> The above setupdata block introduction means we need a list of all available animations, with their settings. Likely we want to at least attach a name to make it understandable what animation we're talking about, perhaps even with a one-two line description.
> 
> Such information would be great to provide to designers so they have a clue what to create.
> 
> For graphics set compatibility, we may also want to state required number of frames (or length in time perhaps) as well as size, people involved, relative start positions, and/or end-positions of objects or people, etc.
> 
> Maybe the current system already solves that in some way?

**Comment by Alberth289346** [comment]
Next to the technical implementation details and requirements to the graphics, we also need a list of images and animations for a complete replacement set. This should be detailed and at the visual level. Not "walk animations for patient", but what direction for which patient variations.

More technical information such as start frame number, number of frames, flags, and layers may be added as a reference to the current graphics set.
